Covariant Returns Feature

fadimounir opened the pull request:
Allowing return types on MethodImpl to be derived types of the return type on the MethodDecl.

Feature limitations:
  - Only supports MethodImpls and MethodDecls on classes (no interfaces/valuetypes).
  - MethodImpl and MethodDecl cannot be on the same type.

MethodImpls where the signature matches for all parameters except for the return type are conditionally treated as valid during enumerating/processing, but flagged for further return type checking at a later stage of type load.

For every type being loaded LOAD_EXACTPARENTS stage, check every flagged overriding method for return type compatibility with the overriden method (using `CanCastTo()`), and process the MethodImpls with the RequireMethodImplToRemainInEffect attribute.

Inludes unit tests to cover positive/negative scenarios:
  - Non-generics
  - Generics, with various levels of complexities on generic instantiations and substitutions
  - GVMs
  - Tests for validation of the existence/absence of the RequireMethodImplToRemainInEffect attribute

**TODOs:**
- [x] Rename attribute to `RequireMethodImplToRemainInEffectAttribute` and add it to System.Runtime reference assembly (and update tests to reference it from there instead of S.P.C)
- [x] Crossgen2 work to support feature (some unit tests are failing today with cg2)
- [x] Add unit tests to cover interface -> interface compatibility
- [x] Add unit tests to cover class -> interface compatibility
- [x] Add unit tests to cover delegate invokes using base type/derived type signatures
- [x] Add coverage for covariant/contravariant generic interfaces used as covariant return types (and verify it works with crossgen2 as well)
- [x] Add unit test to ensure struct -> interface casts are treated as incompatible
- [ ] Visit the uncommon castable scenarios and determine if they should be supported, and add test coverage (pointers, byrefs, arrays->IList/ICollection..., int[]->uint[], etc...)
- [ ] Add a chapter on this to https://github.com/dotnet/runtime/blob/master/docs/design/specs/Ecma-335-Augments.md

cc @dotnet/crossgen-contrib

Commit:
Covariant Returns Feature: Allowing return types on MethodImpl to be derived types of the return type on the MethodDecl.

Feature limitations:
  - Only supports MethodImpls and MethodDecls on classes (no interfaces/valuetypes).
  - MethodImpl and MethodDecl cannot be on the same type.
  - Interface/valuetypes not supported as covariant return types.

Changes are mostly a boolean flag being passed around to allow for covariant type checking in method signatures.
Generics are handled by correctly keeping track of the substitution chain while traversing the base type hierarchy for type comparison.

All method signature comparisons uses metadata for checking, without loading any type (with some exceptions).

Validation for the `ValidateMethodImplRemainsInEffectAttribute` is performed at the very last step of `CLASS_LOAD_EXACTPARENTS`.

Inludes unit tests to cover positive/negative scenarios:
  - Non-generics
  - Generics, with various levels of complexities on generic instantiations and substitutions
  - GVMs
  - Delegates
  - Tests for validation of the existance/absence of the ValidateMethodImplRemainsInEffectAttribute
  - Interface and valuetype scenarios (negative scenarios)

fadimounir commented:
Posting this to start getting feedback while working on the Crossgen2 changes.

@davidwrighton @jkotas Who can I talk to regarding the Mono side of the work?

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/ValidateMethodImplRemainsInEffectAttribute.cs:
```diff
@@ -0,0 +1,11 @@
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
+    public sealed class ValidateMethodImplRemainsInEffectAttribute : Attribute
```
This should be run through API review.

Also, should the attribute be used in any of the tests?

fadimounir replied:
Who should I talk to for the API review?

> Also, should the attribute be used in any of the tests?

Yes, there are some tests with invalid return types that expect the presence of this attribute to cause a TypeLoadException to be thrown.

jkotas replied:
> Who should I talk to for the API review?

Look for recent email from Immo titled "API Review"

fadimounir replied:
https://github.com/dotnet/runtime/issues/35315

jkotas commented on src/coreclr/tests/src/Loader/classloader/MethodImpl/CovariantReturns/Validation/OverrideSameSigAsDecl.il:
```diff
@@ -0,0 +1,483 @@
+
+.class public auto ansi beforefieldinit Invalid2<U,V> extends class GenTestType<!U,!V>
+{
+  .custom instance void [System.Private.CoreLib]System.Runtime.CompilerServices.ValidateMethodImplRemainsInEffectAttribute::.ctor() = (01 00 00 00)
```
This should be added to `System.Runtime` ref assembly and referenced via `[System.Runtime]`. We should avoid adding tests that depend on `System.Private.CoreLib` directly. They cause a lot of complications.

fadimounir replied:
How can I add it to the `System.Runtime` facade?

jkotas replied:
Add it to https://github.com/dotnet/runtime/tree/master/src/libraries/System.Runtime/ref

jkotas commented on docs/design/features/covariant-return-methods.md:
```diff
@@ -0,0 +1,83 @@
+# Covariant Return Methods
+
+Covariant return methods is a runtime feature designed to support the [covariant return types](https://github.com/dotnet/csharplang/blob/master/proposals/covariant-returns.md) and [records](https://github.com/dotnet/csharplang/blob/master/proposals/records.md) C# language features posed for C# 9.0.
```
Could you please add a chapter on this to https://github.com/dotnet/runtime/blob/master/docs/design/specs/Ecma-335-Augments.md ? Try to match the style and language used by ECMA-335 (e.g. in the ideal case - we would just copy&paste it over once we figure out how to make edits to ECMA-335).

jkotas commented on src/coreclr/src/vm/siginfo.cpp:
```diff
@@ -3567,6 +3567,346 @@ BOOL CompareTypeTokens(mdToken tk1, mdToken tk2, Module *pModule1, Module *pModu
+}
+
+//------------------------------------------------------------------
+// Extract the base type's signature and stubstitution from the second type signature
```
```suggestion
// Extract the base type's signature and substitution from the second type signature
```

fadimounir replied:
Fixed

jkotas commented on src/coreclr/src/vm/methodtablebuilder.cpp:
```diff
@@ -2365,6 +2444,9 @@ MethodTableBuilder::EnumerateMethodImpls()
                     BuildMethodTableThrowException(IDS_CLASSLOAD_MI_MISSING_SIG_BODY);
                 }
+
+                BOOL allowCovariantReturn = IsEligibleForCovariantReturns(theDecl);
```
Would it make sense to do this only if the ordinary `CompareMethodSigs` fails?

fadimounir replied:
It's possible yes. I guess that actually makes more sense, since probably 90% of the cases won't have covariant return types, so we can skip eligibility checking here.

jkotas commented on docs/design/features/covariant-return-methods.md:
```diff
@@ -0,0 +1,83 @@
+
+For virtual method slot MethodImpl overrides, each slot shall be checked for compatible signature on type load. (Implementation note: This behavior can be triggered only if the type has a covariant return type override in its hierarchy, so as to make this pay for play.)
+
+A new `ValidateMethodImplRemainsInEffectAttribute` shall be added. The presence of this attribute is to require the type loader to ensure that the MethodImpl records specified on the method have not lost their slot unifying behavior due to other actions. In other words, when a MethodImpl on type A overrides some method using a derived return type in the signature, any type deriving from A will be ...
```
This says the attribute is placed on methods, but the implementation places it on types.

fadimounir replied:
Ah yes, thanks for catching that. I read these specs a while back, so I did the implementation based on memory :). I'll fix that

gafter replied:
I don't know at this point whether the requirement is to place it on the method or on the class. I need to know that for the compiler.

---
In reply to: [413348440](https://github.com/dotnet/runtime/pull/35308#discussion_r413348440) [](ancestors = 413348440)

jkotas commented on src/coreclr/src/vm/siginfo.cpp:
```diff
@@ -3567,6 +3567,346 @@ BOOL CompareTypeTokens(mdToken tk1, mdToken tk2, Module *pModule1, Module *pModu
+        PRECONDITION(CheckPointer(pBaseTypeDefOrRefToken));
+        PRECONDITION(CheckPointer(ppBaseTypeTokenModule));
+        PRECONDITION(TypeFromToken(tk) == mdtTypeRef || TypeFromToken(tk) == mdtTypeDef);
+        MODE_ANY;
```
It feels dangerous to run all this with `MODE_ANY`. Can it be restricted to `MORE_PREEMPTIVE` ?

fadimounir replied:
The main comparison algorithm is implemented as MODE_ANY (`CompareElementType()`). Would there be any side effect from making the new APIs MODE_PREEMPTIVE, while the caller `CompareElementType()` is MODE_ANY?

jkotas replied:
Does that main comparison algorithm need to be `MODE_ANY` ? Can we flip the whole thing to MODE_PREEMPTIVE ?

fadimounir replied:
I can't say with certainty. Maybe @davidwrighton can help here

jkotas commented:
> Who can I talk to regarding the Mono side of the work?

@marek-safar ?

We should have the Mono side figured out before starting the work next time.

Commit:
Adding the missing .ctor methods to the various types

fadimounir commented:
@davidwrighton @dotnet/crossgen-contrib I tested the crossgen2 behavior, and it is currently doing the right thing. There are no incorrect devirtualizations happening:
``` asm
call qword ptr [0x6138]       // GenTestType<int, object, int, object> (NEW_OBJECT)
mov rcx, rax
lea rdx, [rsp + 32]
lea r11, [0x60c0]             // GenRetType<!0,Dictionary<String,!1>> GenBaseType<int, object>.MyGenFunc(ref String) (VIRTUAL_ENTRY)
cmp dword ptr [rcx], ecx
call qword ptr [0x60c0]       // GenRetType<!0,Dictionary<String,!1>> GenBaseType<int, object>.MyGenFunc(ref String) (VIRTUAL_ENTRY)
```

The loading of the `VIRTUAL_ENTRY` and of the type's `MethodTable` and vtable layout are handled correctly by the runtime at this point.

Unless there is something I am missing, it seems like we can consider the crossgen2 work here as completed. I don't see the need to implement any of this in the crossgen2's TypeSystem. Thoughts?

Commit:
Perform covariant return signature matching only after an exact match fails

Fix a couple of issues in IL tests

marek-safar commented:
> We should have the Mono side figured out before starting the work next time.

I agree such change has to be coordinated from the beginning as we now need to re-prioritize other work otherwise we ship with incompatible runtimes for net5.   Please loop @lambdageek and @vargaz on type-system design decisions from the beginning.

Commit:
Add test coverage for implicit override with less derived return type

Commit:
Change ValidateMethodImplRemainsInEffect attribute to apply to methods

fadimounir commented:
The validation around the `ValidateMethodImplRemainsInEffectAttribute` is a bit incorrect here. My understanding was a bit incorrect. I'm reworking this part now and updating the unit tests to verify the correct behavior

Commit:
Moving covariant return type checking to the final stage of type loading, and use CanCastTo instead of signature-based checking, to be allow for type compatibility based on ECMA I.8.7.1.

Fixed behavior of the ValidateMethodImplRemainsInEffectAttribute.

Updates to tests

Commit:
Rename attribute and reference it from System.Runtime instead of S.P.C

Commit:
Add test coverage for interface cases.

Fixed an assert

gafter commented on docs/design/features/covariant-return-methods.md:
```diff
@@ -0,0 +1,96 @@
+
+Covariant return methods is a runtime feature designed to support the [covariant return types](https://github.com/dotnet/csharplang/blob/master/proposals/covariant-returns.md) and [records](https://github.com/dotnet/csharplang/blob/master/proposals/records.md) C# language features posed for C# 9.0.
+
+This feature allows an overriding method to have a return type that is different than the one on the method it overrides, but compatible with it. The type compability rules are defined in ECMA I.8.7.1. Example: using a more derived return type. The only exception applicable to this rule for this feature is the compatibility between a value type and an interface that it implements, and this is due ...
```
>exception [](start = 254, length = 9)

This is not an exception to ECMA I.8.7.1. The conversion from a value type to an interface it implements is not included in the *compatible-with* relation defined there.

fadimounir replied:
Right. This is more of an implementation detail with the use of `CanCastTo()`. I'll reword it.

gafter replied:
Honestly, I prefer the reference to the specification. You're welcome to also describe the implementation strategy, but to author metadata/IL we need to know the specification.

---
In reply to: [418240009](https://github.com/dotnet/runtime/pull/35308#discussion_r418240009) [](ancestors = 418240009)

Commit:
Handling for struct cases, and adding test coverage for it.

Commit:
Small test fixes

Commit:
Fix consistency issue with the way the RequireMethodImplToRemainInEffectAttribute is declared

Commit:
Support covariant returns and slot unifications in crossgen2

These changes fix the behavior of the devirtualization algorithm to not incorrectly devirtualize covariant return methods that have the slot unification attribute

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs:
```diff
@@ -99,6 +102,42 @@ public Enumerator GetEnumerator()
                 return new Enumerator(_members);
             }
 
+            public IEnumerable<MethodDesc> MethodsRequiringSlotUnification
```
Make this no allocation by doing this:

```suggestion
            public Enumerable MethodsRequiringSlotUnification => new Enumerable(_methodsRequiringSlotUnification);
```

With a matching:

```csharp
public readonly struct Enumerable
{
    private readonly MethodDesc[] _arrayToEnumerate;

    public Enumerable(MethodDesc[] arrayToEnumerate)
    {
        _arrayToEnumerate = arrayToEnumerate;
    }

    public Enumerator GetEnumerator()
    {
        return new Enumerator(_arrayToEnumerate);
    }
}
```

MichalStrehovsky reviewed (commented):
Not actually looking at the resolution logic in detail because I'm not up to speed on the unification group business anyway.

Also, do we need a `ValidateMethodsWithCovariantReturnTypes` equivalent in crossgen2?

lambdageek commented on src/coreclr/src/tools/Common/TypeSystem/Common/MethodDesc.cs:
```diff
@@ -114,7 +114,7 @@ public int Length
         }
 
-        public bool Equals(MethodSignature otherSignature)
+        public bool Equals(MethodSignature otherSignature, bool allowCovariantReturn = false)
```
Agree with Michal.  Equality is a symmetric relation (`x.Equals (y)` and `y.Equals(x)` should both return the same value), adding a boolean flag that breaks symmetry is surprising.

In type theory, the relation between two types (in this case two function signatures) where one is more general than the other (for example because the return type of one function is a supertype of the return type of the other) is called "subsumption".  That might be a good name to use here.

fadimounir replied:
Sounds reasonable yes.

MichalStrehovsky commented:
Since this is using `CanCastTo`... how do we feel about the fact that this is going to enable things like:

* override a method returning `int[]` with a method returning `uint[]`, or
* for bonus points even this:

```csharp
class Base
{
    public virtual IntPtr[] Fun() => null;
}

// This is obvious pseude-code because C# won't let us introduce methods differing
// in return type. C# also requires to be explicit about "virtual" and "override".
// But IL... not so much.
class Derived
{
    // overrides Base.Fun on 32bit platforms.
    public virtual uint[] Fun() => null;

    // overrides Base.Fun on 64bit platforms.
    public virtual ulong[] Fun() => null;
}
```

janvorli commented on src/coreclr/src/tools/Common/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs:
```diff
@@ -99,6 +102,42 @@ public Enumerator GetEnumerator()
+                {
+                    foreach (var method in _methodsRequiringSlotUnification)
+                    {
+                        if (method == null)
```
How could `null` appear in the _methodsRequiringSlotUnification?

MichalStrehovsky replied:
It's `Array.Resize`d a couple lines down - null entries mark unused space left from that.

Maybe this could use an `ArrayBuilder<MethodDesc>` to wrap the detail that the array is only usable up to `_methodsRequiringSlotUnificationCount`.

fadimounir commented on src/coreclr/src/tools/Common/TypeSystem/Common/MethodDesc.ToString.cs:
```diff
@@ -46,7 +46,9 @@ public override string ToString()
             var sb = new StringBuilder();
 
-            // (Skipping return type to keep things short)
+            sb.Append(Signature.ReturnType);
```
Ok I can revert that. I was hoping this would provide better diagnostics, not just for covariant return methods, but also for cases where we could have methods with the same name/sig, but different return types.

fadimounir commented on src/coreclr/src/tools/Common/TypeSystem/Common/MethodDesc.cs:
```diff
@@ -124,7 +124,18 @@ public bool Equals(MethodSignature otherSignature)
+                if (!allowCovariantReturn)
+                    return false;
+
+                // Casting a struct to an interface it implements is allowed, but these two types are not compatible and
```
The feature should be able to extend to something like pointers (see ECMA 8.7.2), but these are all interesting corner cases that we should probably call out and clarify what is supported/not-supported, and see if we need to special case some of them from the `CanCastTo()` behavior (ex: valuetype to interface casting, where `CanCastTo()` in the core TypeSystem does not assume boxed sources, unlike the implementation here)

MichalStrehovsky replied:
Should we add the ref return/pointer return cases to the TODO above so that it doesn't get lost?

fadimounir commented on src/coreclr/src/tools/Common/TypeSystem/Common/TypeSystemHelpers.cs:
```diff
@@ -411,5 +411,29 @@ public static bool ContainsSignatureVariables(this TypeDesc thisType)
+        /// </summary>
+        /// <param name="method">Method to check</param>
+        /// <returns>True when the method is a MethodImpl marked with the RequireMethodImplToRemainInEffect custom attribute, false otherwise.</returns>
+        public static bool RequiresSlotUnification(this MethodDesc method)
```
I have not measured perf

fadimounir commented on src/coreclr/src/tools/Common/TypeSystem/Common/TypeSystemHelpers.cs:
```diff
@@ -411,5 +411,29 @@ public static bool ContainsSignatureVariables(this TypeDesc thisType)
+            MetadataType mdType = method.OwningType as MetadataType;
+            if (mdType != null)
+            {
+                foreach (MethodImplRecord methodImplRecord in mdType.VirtualMethodImplsForType)
```
Roslyn will not, but this is about being on par with some of the verification on the coreclr Typesystem side. However, the virtual resolution algorithm should not invoke this for non-MethodImpls, so we can just convert this into a debug-only check/assert, or completely remove it

MichalStrehovsky replied:
My vote would be to remove it then.

fadimounir replied:
I like the idea of removing it, but let's do that closer to merging. It's useful to have this check for now during development

Commit:
some CR feedback

davidwrighton commented:
@MichalStrehovsky Those are all the `compatible-with` concept from the ECMA spec. I'd rather see those oddities appear than miss cases.

davidwrighton commented:
@fadimounir I'd like to see the use of CanCastTo replaced with a helper function that represents the `compatible-with` concept, which should make it clear what semantic we're actually enabling here. The CanCastTo usage needs to be carefully held back to only cover the cases where using the casting helper is the right choice.

fadimounir commented:
@davidwrighton Thinking a bit about `CanCastTo` vs `compatible-with`, I'm trying to understand why we would want to focus on `compatible-with` and come up with a new helper instead of using `CanCastTo`? The way I look at it, if an overriding method can return an object of a type that can be casted to the return type of the overriden method, then things will work correctly. There are only a handful of cases where we can cast from the return type of the overriding method to the return type of the overriden method, and vice versa. I believe most cases will involve a more derived class/interface being returned from the overriding method, which can only be a one-way cast.

We can spend time implementing a new helper, but there is some risk involved there in getting all the cases we care about to work correctly (variance, equivalent interfaces, interfaces, arrays, pointers, byrefs, etc...). Using `CanCastTo` would have the advantage of leveraging code that has been thoroughly tested in the past.

Am I missing something?

MichalStrehovsky commented:
> @MichalStrehovsky Those are all the `compatible-with` concept from the ECMA spec. I'd rather see those oddities appear than miss cases.

Right, but if we're conscious about supporting that, we should have test coverage.

AlekseyTs commented on docs/design/features/covariant-return-methods.md:
```diff
@@ -0,0 +1,98 @@
+
+MethodImpl checking will allow a return type to vary as long as the override is compatible with the return type of the method overriden (ECMA I.8.7.1).
+
+If a language wishes for the override to be semantically visible such that users of the more derived type may rely on the covariant return type it shall make the override a newslot method with appropriate visibility AND name to be used outside of the class.
```
This statement sounds like newslot  is not required if the method is not semantically visible (BTW, what exactly does it mean to be semantically visible). Is this correct?

davidwrighton commented:
@fadimounir I'm not saying that we shouldn't use CanCastTo, but what I'm saying is that we should have a helper which is implemented (probably with CanCastTo internally for most of its logic) that expresses the compatible-with relationship. Michal has called out a few cases where we need to be precise and not just blindly call CanCastTo, and I think that list may grow to 2-3 items total before we're done.

Commit:
Add unit test coverage for delegates

marek-safar commented:
/cc @joshpeterson

janvorli committed:
Fix handling of covariant and contravariant generic interfaces

Also add tests for those.

janvorli commented:
@davidwrighton I've added a commit that fixes handling of the covariant and contravariant generic interfaces as return values. Can you please take a look?

davidwrighton commented:
@janvorli, 
It looks right to pass NULL to CanCastTo as the last argument.
For the load level assert change, you need to test a case where the generic arg load level could be a problem.

For instance something like the following
```
class Base<T> {}
class Derived : Base<Derived> {}
class Derived2 : Base<Derived2> {}
interface IVariant<out V>
class BaseClassWithVirtualMethod
{
    public virtual IVariant<Base<Derived>> Func() {...}
}

class DerivedClassPass1
{
    public override IVariant<Derived> Func() {} // This should work
}

class DerivedClassFail1
{
    public override IVariant<Derived2> Func() {} // This should fail
}

janvorli committed:
Added test cases to interfaces unit test

Based on PR feedback, I've added test cases to verify the load level
asserts change.

janvorli commented:
@davidwrighton I've added a commit with a test based on your suggestion and tested it with debug build of runtime. It works as expected.

davidwrighton commented:
The generic variance tests look good. Looking through the tests I think we need to have Nullable tests as well to ensure that CanCastTo isn't causing a problem. If you want to be particular, testing to see if int? and int are compatible-with. They should not be for this calculation.

janvorli committed:
Fix Nullable handling and add more tests

The attempt to overload method with int32 return value by a method with
Nullable<int32> was asserting in CanCastTo. This change fixes it by
pre-checking this case before calling into CanCastTo.
I've also slightly refactored the
ValidateMethodsWithCovariantReturnTypes by extracting the
compatible-with stuff into IsCompatibleWith method.

Added tests to test various compatible-with differences from the
originally used CanCastTo

janvorli commented:
@davidwrighton I've tested and fixed the nullable case and refactored the code a tiny bit by introducing IsCompatibleWith method that wraps around the CanCastTo and does some of the pre-filtering as you've suggested.
I've added another set of tests for testing various interesting cases from the compatible-with list, like overriding return value of IList<int32> by uint32[] etc.

davidwrighton commented:
Have we gotten api signoff on RequireMethodImplToRemainInEffectAttribute yet? Otherwise I think this is implemented and the logic looks good.

janvorli commented:
> Have we gotten api signoff on RequireMethodImplToRemainInEffectAttribute yet

Doesn't seem so, #35315.

fadimounir commented on src/coreclr/src/vm/class.cpp:
```diff
@@ -1067,10 +1090,7 @@ void ClassLoader::ValidateMethodsWithCovariantReturnTypes(MethodTable* pMT)
         {
             GCX_COOP();
```
It might be better to move GCX_COOP inside the IsCompatibleWith helper when needed (not all checks need this)

janvorli replied:
Thanks Fadi, it makes sense. I've made that change. Btw, I am still using your branch for finishing this work, so I'd like to ask you to keep it until it is merged.

fadimounir replied:
Thanks Jan, yes I’ll keep the branch around

janvorli committed:
Reflect Fadi's feedback - limit the GCX_COOP scope

MichalStrehovsky commented:
> refactored the code a tiny bit by introducing IsCompatibleWith method that wraps around the CanCastTo and does some of the pre-filtering as you've suggested.

Do we need to port that to crossgen2 as well?

In general I don't believe the CompatibleWithTest.il test that tests the CanCastTo pre-filtering is going to work with crossgen2 and optimizations enabled - I expect RyuJIT to optimize out the `newobj/pop` sequence and since crossgen2 doesn't actually "load" types, it won't see there's a problem.

It might be better to do newobj/callvirt to ensure the virtual method resolution in crossgen2 is actually hit and we test what we want to test.

janvorli commented:
> In general I don't believe the CompatibleWithTest.il test that tests the CanCastTo pre-filtering is going to work with crossgen2 and optimizations enabled - I expect RyuJIT to optimize out the `newobj/pop` sequence and since crossgen2 doesn't actually "load" types, it won't see there's a problem.

I've tested it with crossgen2 and optimizations enabled and the test was passing (that means the parts that should succeed were succeeding and the parts that should fail were failing).  But it makes sense to change the newobj/pop to newobj / callvirt.

> Do we need to port that to crossgen2 as well?

Right, I've missed that

janvorli committed:
Reflect PR feedback

- Change the CompatibleWithTest to call virtual methods on test
classes instead of just instantiating the classes.
- Update crossgen2 similar to the runtime to have IsCompatibleWith
method doing the same pre-filtering before calling CanCastTo.

janvorli commented:
@MichalStrehovsky I've made changes based on your feedback.

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs:
```diff
@@ -20,6 +20,31 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
+
+        private static bool IsEquivalentTo(this TypeDesc thisType, TypeDesc otherType)
+        {
+            // TODO: Once type equivalence is implemented, this implementation needs to be enhanced to honor it.
```
Is type equivalence used for anything outside COM?

We don't intend to model COM in the managed type system so if this is exclusive to COM, the use of this can be replaced with `==` (we can safely use refence equality in the type system; no need to call `Equals`).

janvorli replied:
I think that it is used just by COM, @davidwrighton can you please confirm that?

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/CastingHelper.cs:
```diff
@@ -20,6 +20,31 @@ public static bool CanCastTo(this TypeDesc thisType, TypeDesc otherType)
+                return false;
+            }
+
+            return otherType.CanCastTo(thisType);
```
This is still going to throw if thisType or otherType is a pointer or function pointer. Do we need to handle that? (I had that feedback in https://github.com/dotnet/runtime/pull/35308#discussion_r420007424 but it probably got lost.

janvorli replied:
Ah, let me add those to the test suite. It seems pointers cannot be compatible-with each other unless they are the same. I am actually not sure if the function pointers correspond to "method-signature compatible-with" case in the ECMA spec.

MichalStrehovsky reviewed (commented):
Thanks!

janvorli committed:
Rename the attribute to PreserveBaseOverridesAttribute

janvorli committed:
Disable covariant returns tests on Mono for now

The mono doesn't implement the feature yet

davidwrighton reviewed (approved):
This looks good now. I believe we may have some small amount of cleanup around error conditions about pointers and function pointers, but we should handle those as bugs.

MichalStrehovsky commented:
Are we happy about the perf of virtual method resolution after this change? https://github.com/dotnet/runtime/pull/35308#discussion_r420018828

It feels like this attribute should be cached the same way we cache e.g. `UnmanagedCallersOnlyAttribute`...

davidwrighton commented:
While not great, and I agree we should probably cache it, the perf impact isn't that significant in crossgen2 due to the relatively low amount of virtual resolution it performs. We'll probably need to adjust that when we port this to the CoreRT codebase where virtual resolution performance is much more important. However, that is bug fix level work that doesn't need to block checkin.

janvorli commented:
I am going to create an issue tracking all the additional changes that were mentioned here so that they don't get lost.
